pkg/generator/dag_generation.go

```diff
@@ -0,0 +1,233 @@
+				buffer += "     "
+			}
+			printMessage = buffer + printMessage
+			startingNode = false
+		} else {
+			printMessage = printMessage + " -> " + strconv.Itoa(functionId)
+		}
+		for i := 0; i < len(DAGNode.Value.(*common.Node).Branches); i++ {
+			nodeQueue = append(nodeQueue, dummyNode)
+			copy(nodeQueue[1:], nodeQueue)
+			nodeQueue[0] = DAGNode.Value.(*common.Node).Branches[i].Front()
+		}
+		if DAGNode.Next() == nil {
+			println(printMessage)
+			buffer = ""
+			if len(nodeQueue) > 0 {
+				startingNode = true
+			} else {
+				break
+			}
+		} else {
+			nodeQueue = append(nodeQueue, dummyNode)
+			copy(nodeQueue[1:], nodeQueue)
+			nodeQueue[0] = DAGNode.Next()
+		}
+	}
+}
+
+// Read the Cumulative Distribution Frequency (CDF) of the widths and depths of a DAG
+func generateCDF(file string) [][]float64 {
+	f, err := os.Open(file)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer f.Close()
+	csvReader := csv.NewReader(f)
+	records, err := csvReader.ReadAll()
+	if err != nil {
+		log.Fatal(err)
+	}
+	records = records[1:]
+	cdf := make([][]float64, len(records[0]))
+	for i := 0; i < len(records[0]); i++ {
+		cdf[i] = make([]float64, len(records))
+	}
+	for i := 0; i < len(records[0]); i += 2 {
+		for j := 0; j < len(records); j++ {
+			cdfProb, _ := strconv.ParseFloat(strings.TrimSuffix(records[j][i+1], "%"), 64)
+			cdfValue, _ := strconv.ParseFloat(records[j][i], 64)
+			cdf[i+1][j] = cdfProb
+			cdf[i][j] = cdfValue
+			if cdfProb == 100.00 {
+				cdf[i] = cdf[i][:j+1]
+				cdf[i+1] = cdf[i+1][:j+1]
+				break
+			}
+		}
+	}
+	return cdf
+}
+
+// Generate pseudo-random probabilities and compare it with the given CDF to obtain the depth and width of the DAG
+func getDAGStats(cdf [][]float64, maxSize int, numberOfTries int) (int, int) {
+	var width, depth int
+	depthProb := rand.Float64() * 100
+	widthProb := rand.Float64() * 100
+	for i, value := range cdf[1] {
+		if value >= widthProb {
+			width = int(cdf[0][i])
+			break
+		}
+	}
+	for i, value := range cdf[3] {
+		if value >= depthProb {
+			depth = int(cdf[2][i])
+			break
+		}
+	}
+	// Re-run DAG Generation if size exceeds number of functions
+	if maxSize < width*(depth-1)+1 {
+		if numberOfTries == 10 {
+			return 1, maxSize
+		}
+		width, depth = getDAGStats(cdf, maxSize, numberOfTries+1)
+	}
+	return width, depth
+}
+
+func GetMaxInvocation(functionList []*common.Function) []int {
+	maxInvocation := make([]int, len(functionList[0].InvocationStats.Invocations))
+	for _, i := range functionList {
+		for index, invocation := range i.InvocationStats.Invocations {
+			maxInvocation[index] = max(maxInvocation[index], invocation)
+		}
+	}
+	return maxInvocation
+}
+
+func GenerateDAG(config *config.LoaderConfiguration,functions []*common.Function) []*list.List{
+	var width,depth int
+	DAGDistribution := generateCDF(fmt.Sprintf("%s/dag_structure.csv", config.TracePath))
+	totalLinkedList := []*list.List{}
+	for _, function := range functions {
```

Comment:
Noted, after discussing with Hancheng, I will change the workflow to generate a single DAG only, based on the given parameters. And add a parameter to determine the entry function in the config file.

Reply:
Why only one? IIRC, the DAGs are quite small on average. So even if we have 2k functions deployed, it would use only a handful. Can you continue to generate DAGs until it won't be able to fit another one? That seems more appropriate.

Reply:
Thanks for the suggestion. Have made it continuously generate DAGs. The entry function will start at index 0, with the next DAG's entry function being the next unused function. (ie: DAG 1 uses f(0) -> f(1) ->f(2). DAG 2 will have entry function f(3)). Invocation Frequencies and IAT will follow the respective entry function. DAGs will have different shapes based on dag_structure.csv unless user specifies with `EnableDAGDataset`.